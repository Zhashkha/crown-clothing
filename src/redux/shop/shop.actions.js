import ShopActionTypes from "./shop.types";

import {
  firestore,
  convertCollectionsSnapshotToMap
} from "../../firebase/firebase.utils";

export const fetchCollectionsStart = () => ({
  type: ShopActionTypes.FETCH_COLLECTIONS_START
});

export const fetchCollectionsStartAsync = () => {
  return (dispatch) => {
    dispatch(fetchCollectionsStart());
    const collectionRef = firestore.collection("collections");

    collectionRef
      .get()
      .then((snapshot) => {
        const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
        dispatch(fetchCollectionSuccess(collectionsMap));
      })
      .catch((error) => dispatch(fetchCollectionFailure(error.message)));
  };
};

export const fetchCollectionSuccess = (collectionsMap) => ({
  type: ShopActionTypes.FETCH_COLLECTIONS_SUCCESS,
  payload: collectionsMap
});

export const fetchCollectionFailure = (errorMessage) => ({
  type: ShopActionTypes.FETCH_COLLECTIONS_FAILURE,
  payload: errorMessage
});
