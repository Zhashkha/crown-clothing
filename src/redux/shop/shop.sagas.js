import { takeLatest, put, all, call } from "redux-saga/effects";

import ShopActionTypes from "./shop.types";
import {
  firestore,
  convertCollectionsSnapshotToMap
} from "../../firebase/firebase.utils";
import { fetchCollectionSuccess, fetchCollectionFailure } from "./shop.actions";

export function* fetchCollectionsStartAsync() {
  try {
    const collectionRef = firestore.collection("collections");
    const snapshot = yield call([collectionRef, "get"]);
    const collectionsMap = yield call(
      convertCollectionsSnapshotToMap,
      snapshot
    );
    yield put(fetchCollectionSuccess(collectionsMap));
  } catch (error) {
    yield put(fetchCollectionFailure(error.message));
  }
}

export function* onFetchCollectionsStart() {
  yield takeLatest(
    ShopActionTypes.FETCH_COLLECTIONS_START,
    fetchCollectionsStartAsync
  );
}

export function* shopSagas() {
  yield all([call(onFetchCollectionsStart)]);
}
