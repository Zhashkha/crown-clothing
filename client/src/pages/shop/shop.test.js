import React from "react";
import { mount } from "enzyme";
import { combineReducers, createStore } from "redux";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";

import { ShopPage } from "./shop.component";

const createMockStore = ({ state, reducers }) => {
  const store = createStore(combineReducers(reducers), state);

  return {
    ...store,
    persistor: {
      persist: () => null
    }
  };
};

describe("ShopPage", () => {
  let wrapper;
  let store;
  let mockFetchCollectionsStart;

  beforeEach(() => {
    const mockReducer = (state = { isFetching: true }, action) => state;

    const mockState = {
      shop: {
        isFetching: true
      }
    };

    store = createMockStore({
      state: mockState,
      reducers: { shop: mockReducer }
    });

    const mockMatch = {
      path: ""
    };

    mockFetchCollectionsStart = jest.fn();

    const mockProps = {
      match: mockMatch,
      fetchCollectionsStart: mockFetchCollectionsStart
    };

    wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <ShopPage {...mockProps} />
        </BrowserRouter>
      </Provider>
    );
  });

  it("test ShopPage snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("ShopPage should be mounted and therefore fetchCollectionsStart should be called", () => {
    expect(mockFetchCollectionsStart).toHaveBeenCalled();
  });
});
