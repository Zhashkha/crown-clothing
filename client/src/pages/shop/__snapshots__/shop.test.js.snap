// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage test ShopPage snapshot 1`] = `ReactWrapper {}`;
